# Análisis Exploratorio de Datos (EDA)

Este cuaderno realiza un EDA básico sobre un dataset hipotético de compras. El análisis se enfoca en preguntas de negocio sencillas y aplica métodos de filtrado y agrupación para calcular métricas relevantes.

In [3]:
!wget https://raw.githubusercontent.com/javierherrera1996/IntroMachineLearning/refs/heads/main/PrimerCorte/RegresionLineal/shopping_trends.csv

--2025-02-27 22:15:56--  https://raw.githubusercontent.com/javierherrera1996/IntroMachineLearning/refs/heads/main/PrimerCorte/RegresionLineal/shopping_trends.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.109.133, 185.199.111.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 453248 (443K) [text/plain]
Saving to: ‘shopping_trends.csv.1’

shopping_trends.csv 100%[===================>] 442.62K  --.-KB/s    in 0.05s   

2025-02-27 22:15:56 (9.42 MB/s) - ‘shopping_trends.csv.1’ saved [453248/453248]



## 1. Importar librerías y cargar datos

En este ejemplo, se asume que el archivo CSV se llama `datos_compras.csv`. Ajusta el nombre o la ruta del archivo según corresponda.

In [5]:
import pandas as pd

# Cargar el dataset
df = pd.read_csv('datos_compras.csv')

# Mostrar las primeras filas para verificar
df.head()

Customer ID  Age Gender Item Purchased  Category  Purchase Amount (USD)  \
0            1   55   Male         Blouse  Clothing                     53   
1            2   19   Male        Sweater  Clothing                     64   
2            3   50   Male          Jeans  Clothing                     73   
3            4   21   Male        Sandals  Footwear                     90   
4            5   45   Male         Blouse  Clothing                     49   

        Location Size      Color  Season  Review Rating Subscription Status  \
0       Kentucky    L       Gray  Winter            3.1                 Yes   
1          Maine    L     Maroon  Winter            3.1                 Yes   
2  Massachusetts    S     Maroon  Spring            3.1                 Yes   
3   Rhode Island    M     Maroon  Spring            3.5                 Yes   
4         Oregon    M  Turquoise  Spring            2.7                 Yes   

  Payment Method  Shipping Type Discount Applied Promo Code Used  \
0    Credit Card        Express              Yes             Yes   
1  Bank Transfer        Express              Yes             Yes   
2           Cash  Free Shipping              Yes             Yes   
3         PayPal   Next Day Air              Yes             Yes   
4           Cash  Free Shipping              Yes             Yes   

   Previous Purchases Preferred Payment Method Frequency of Purchases  
0                  14                    Venmo            Fortnightly  
1                   2                     Cash            Fortnightly  
2                  23              Credit Card                 Weekly  
3                  49                   PayPal                 Weekly  
4                  31                   PayPal               Annually

Observamos que nuestro DataFrame contiene, entre otras, las siguientes columnas:

- **Customer ID**: ID único de cada cliente.
- **Age**: Edad del cliente.
- **Gender**: Género del cliente (Male/Female).
- **Item Purchased**: Nombre del producto comprado.
- **Category**: Categoría del producto (Footwear, Clothing, Accessories, etc.).
- **Purchase Amount (USD)**: Monto de la compra en dólares.
- **Location**: Ubicación/Estado.
- **Size**: Talla del producto.
- **Color**: Color del producto.
- **Season**: Temporada (Summer, Winter, etc.).
- **Review Rating**: Calificación de reseña (escala 1-5).
- **Subscription Status**: Indica si el cliente está suscrito (Yes/No).
- **Shipping Type**: Tipo de envío (Express, Store Pickup, etc.).
- **Discount Applied**: Indica si se aplicó descuento (Yes/No).
- **Promo Code Used**: Indica si se usó código promocional (Yes/No).
- **Previous Purchases**: Número de compras anteriores.
- **Payment Method**: Método de pago (Credit Card, Debit Card, PayPal, etc.).
- **Frequency of Purchases**: Frecuencia de compra (Daily, Weekly, Monthly, etc.).


## 2. Limpieza y revisión rápida de datos
Antes de responder a las preguntas de negocio, revisemos datos básicos: número de filas, columnas y valores nulos.

In [6]:
# Ver información general
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3900 entries, 0 to 3899
Data columns (total 19 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Customer ID               3900 non-null   int64  
 1   Age                       3900 non-null   int64  
 2   Gender                    3900 non-null   object 
 3   Item Purchased            3900 non-null   object 
 4   Category                  3900 non-null   object 
 5   Purchase Amount (USD)     3900 non-null   int64  
 6   Location                  3900 non-null   object 
 7   Size                      3900 non-null   object 
 8   Color                     3900 non-null   object 
 9   Season                    3900 non-null   object 
 10  Review Rating             3900 non-null   float64
 11  Subscription Status       3900 non-null   object 
 12  Payment Method            3900 non-null   object 
 13  Shipping Type             3900 non-null   object 
 14  Discount

In [7]:
# Ver estadísticos descriptivos de las columnas numéricas
df.describe()

Customer ID          Age  Purchase Amount (USD)  Review Rating  \
count  3900.000000  3900.000000            3900.000000    3900.000000   
mean   1950.500000    44.068462              59.764359       3.749949   
std    1125.977353    15.207589              23.685392       0.716223   
min       1.000000    18.000000              20.000000       2.500000   
25%     975.750000    31.000000              39.000000       3.100000   
50%    1950.500000    44.000000              60.000000       3.700000   
75%    2925.250000    57.000000              81.000000       4.400000   
max    3900.000000    70.000000             100.000000       5.000000   

       Previous Purchases  
count         3900.000000  
mean            25.351538  
std             14.447125  
min              1.000000  
25%             13.000000  
50%             25.000000  
75%             38.000000  
max             50.000000

Si encuentras valores nulos en alguna columna, podrías decidir cómo manejarlos (eliminarlos o rellenarlos). Por simplicidad, asumimos que no hay muchos datos faltantes o que no es crítico para este ejemplo.

## 3. Preguntas de negocio
A continuación, formulamos algunas preguntas de negocio y utilizamos operaciones de filtrado y/o agrupaciones para responderlas.

### 3.1 ¿Cuántos clientes únicos hay?
Esta pregunta nos ayuda a entender la base total de clientes.

In [10]:
df['Customer ID']

0          1
1          2
2          3
3          4
4          5
        ... 
3895    3896
3896    3897
3897    3898
3898    3899
3899    3900
Name: Customer ID, Length: 3900, dtype: int64

En la base total tenemos 3899 clientes únicos.

### 3.2 ¿Cuál es el gasto promedio por género?
Para entender si hay diferencias en el monto de compra según el género, hacemos un groupby y calculamos la media.

In [14]:
df.groupby("Gender")["Purchase Amount (USD)"].median()

Gender
Female    60.0
Male      60.0
Name: Purchase Amount (USD), dtype: float64

El gasto promedio por género fémenino y másculino es 60.

### 3.3 ¿Cuál es la calificación de reseña promedio por temporada?
Si queremos ver si las reseñas cambian según la temporada, podemos hacer una agrupación por la columna `Season` y calcular la media.

In [19]:
df.groupby("Season")["Review Rating"].median()

Season
Fall      3.7
Spring    3.8
Summer    3.7
Winter    3.7
Name: Review Rating, dtype: float64

La calificación promedio por temporada son las siguientes: otoño: 3.7, primavera: 3.8, verano: 3.7 e Invierno: 3.7.

### 3.4 ¿Cuál es el método de pago más utilizado?
Una forma rápida es usar `value_counts()` en la columna `Payment Method`.

In [20]:
df["Purchase Amount (USD)"].value_counts()

Purchase Amount (USD)
94     62
32     62
36     62
51     61
90     60
       ..
100    36
87     35
49     35
69     34
61     33
Name: count, Length: 81, dtype: int64

### 3.5 ¿Qué categoría se vende más?
Para saber la categoría más popular, también usamos `value_counts()`.

In [25]:
df["Category"].value_counts()

Category
Clothing       1737
Accessories    1240
Footwear        599
Outerwear       324
Name: count, dtype: int64

La categoría que más se vende es la ropa, con 1737 ventas totales.

### 3.6 ¿Cuántos clientes tienen suscripción activa?
Analizamos la columna `Subscription Status`.

In [32]:
df[df["Subscription Status"]=="Yes"]

Customer ID  Age Gender Item Purchased     Category  \
0               1   55   Male         Blouse     Clothing   
1               2   19   Male        Sweater     Clothing   
2               3   50   Male          Jeans     Clothing   
3               4   21   Male        Sandals     Footwear   
4               5   45   Male         Blouse     Clothing   
...           ...  ...    ...            ...          ...   
1048         1049   30   Male          Socks     Clothing   
1049         1050   45   Male     Sunglasses  Accessories   
1050         1051   48   Male       Sneakers     Footwear   
1051         1052   45   Male          Socks     Clothing   
1052         1053   60   Male          Shoes     Footwear   

      Purchase Amount (USD)       Location Size      Color  Season  \
0                        53       Kentucky    L       Gray  Winter   
1                        64          Maine    L     Maroon  Winter   
2                        73  Massachusetts    S     Maroon  Spring   
3                        90   Rhode Island    M     Maroon  Spring   
4                        49         Oregon    M  Turquoise  Spring   
...                     ...            ...  ...        ...     ...   
1048                     77          Maine    L     Yellow    Fall   
1049                     33       Virginia    S     Purple    Fall   
1050                     64        Florida    M      Beige  Winter   
1051                     53           Ohio    M     Indigo    Fall   
1052                     36       Virginia   XL       Pink    Fall   

      Review Rating Subscription Status Payment Method   Shipping Type  \
0               3.1                 Yes    Credit Card         Express   
1               3.1                 Yes  Bank Transfer         Express   
2               3.1                 Yes           Cash   Free Shipping   
3               3.5                 Yes         PayPal    Next Day Air   
4               2.7                 Yes           Cash   Free Shipping   
...             ...                 ...            ...             ...   
1048            2.7                 Yes           Cash        Standard   
1049            2.8                 Yes    Credit Card    Store Pickup   
1050            3.8                 Yes    Credit Card    Next Day Air   
1051            2.6                 Yes         PayPal   Free Shipping   
1052            3.0                 Yes          Venmo  2-Day Shipping   

     Discount Applied Promo Code Used  Previous Purchases  \
0                 Yes             Yes                  14   
1                 Yes             Yes                   2   
2                 Yes             Yes                  23   
3                 Yes             Yes                  49   
4                 Yes             Yes                  31   
...               ...             ...                 ...   
1048              Yes             Yes                  19   
1049              Yes             Yes                  24   
1050              Yes             Yes                  19   
1051              Yes             Yes                  39   
1052              Yes             Yes                   1   

     Preferred Payment Method Frequency of Purchases  
0                       Venmo            Fortnightly  
1                        Cash            Fortnightly  
2                 Credit Card                 Weekly  
3                      PayPal                 Weekly  
4                      PayPal               Annually  
...                       ...                    ...  
1048                    Venmo            Fortnightly  
1049                     Cash              Quarterly  
1050                    Venmo                Monthly  
1051                   PayPal               Annually  
1052              Credit Card            Fortnightly  

[1053 rows x 19 columns]

1052 clientes tienen la suscripción activa.

### 3.7 ¿Cuál es el gasto promedio por categoría?
Para entender mejor en qué categoría gastan más nuestros clientes.

In [35]:
df.groupby("Category")["Purchase Amount (USD)"].mean()

Category
Accessories    59.838710
Clothing       60.025331
Footwear       60.255426
Outerwear      57.172840
Name: Purchase Amount (USD), dtype: float64

El gasto promedio por categoría es: accesorios: 59.83, ropa: 60.02, calzado: 60.25 y ropa de calle: 57.17.

### 3.8 ¿Cuántos clientes usan código promocional?
Podemos filtrar o contar los valores de la columna `Promo Code Used`.

In [37]:
df[df["Promo Code Used"]== "Yes"]

Customer ID  Age Gender Item Purchased     Category  \
0               1   55   Male         Blouse     Clothing   
1               2   19   Male        Sweater     Clothing   
2               3   50   Male          Jeans     Clothing   
3               4   21   Male        Sandals     Footwear   
4               5   45   Male         Blouse     Clothing   
...           ...  ...    ...            ...          ...   
1672         1673   18   Male          Boots     Footwear   
1673         1674   21   Male         Blouse     Clothing   
1674         1675   35   Male           Coat    Outerwear   
1675         1676   35   Male          Pants     Clothing   
1676         1677   27   Male     Sunglasses  Accessories   

      Purchase Amount (USD)        Location Size      Color  Season  \
0                        53        Kentucky    L       Gray  Winter   
1                        64           Maine    L     Maroon  Winter   
2                        73   Massachusetts    S     Maroon  Spring   
3                        90    Rhode Island    M     Maroon  Spring   
4                        49          Oregon    M  Turquoise  Spring   
...                     ...             ...  ...        ...     ...   
1672                     73  South Carolina    L       Gold    Fall   
1673                     62          Hawaii    M     Violet    Fall   
1674                     56          Alaska    M      Peach  Winter   
1675                     90        Colorado    M      Beige  Spring   
1676                     44   West Virginia    M       Cyan  Winter   

      Review Rating Subscription Status Payment Method  Shipping Type  \
0               3.1                 Yes    Credit Card        Express   
1               3.1                 Yes  Bank Transfer        Express   
2               3.1                 Yes           Cash  Free Shipping   
3               3.5                 Yes         PayPal   Next Day Air   
4               2.7                 Yes           Cash  Free Shipping   
...             ...                 ...            ...            ...   
1672            3.8                  No     Debit Card  Free Shipping   
1673            3.4                  No     Debit Card   Next Day Air   
1674            3.2                  No         PayPal  Free Shipping   
1675            4.7                  No          Venmo   Next Day Air   
1676            3.1                  No         PayPal   Next Day Air   

     Discount Applied Promo Code Used  Previous Purchases  \
0                 Yes             Yes                  14   
1                 Yes             Yes                   2   
2                 Yes             Yes                  23   
3                 Yes             Yes                  49   
4                 Yes             Yes                  31   
...               ...             ...                 ...   
1672              Yes             Yes                  15   
1673              Yes             Yes                  49   
1674              Yes             Yes                  10   
1675              Yes             Yes                  21   
1676              Yes             Yes                  34   

     Preferred Payment Method Frequency of Purchases  
0                       Venmo            Fortnightly  
1                        Cash            Fortnightly  
2                 Credit Card                 Weekly  
3                      PayPal                 Weekly  
4                      PayPal               Annually  
...                       ...                    ...  
1672                    Venmo              Quarterly  
1673                    Venmo              Bi-Weekly  
1674               Debit Card            Fortnightly  
1675              Credit Card               Annually  
1676            Bank Transfer                 Weekly  

[1677 rows x 19 columns]

1676 clientes utilizan el código promocional.

## 4. Conclusiones
- **Clientes únicos**: Nos indica cuántas personas diferentes compran en nuestra tienda.
- **Gasto promedio por género**: Sirve para comprender si existe diferencia significativa en el gasto entre hombres y mujeres.
- **Calificación promedio por temporada**: Ayuda a identificar en qué temporada los clientes suelen estar más satisfechos.
- **Métodos de pago preferidos**: Informa sobre la facilidad o preferencia de pago.
- **Categoría más vendida**: Podríamos enfocarnos en promocionarla o analizar por qué se vende más.
- **Suscripciones**: Entender cuántos clientes están suscritos para estrategias de marketing.
- **Uso de códigos promocionales**: Indica si las promociones son efectivas o no.

Este análisis permite identificar patrones y oportunidades para enfocar acciones comerciales y de marketing.